File: 0525_contiguous_array/find_max_length.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

/**
 * Time: O(n)
 * Space: O(n)
 */
int find_max_length(vector<int> &nums)
{
    unordered_map<int, int> indexes;
    indexes[0] = -1;
    int res = 0, tmp = 0;
    for (int i = 0; i < nums.size(); i++)
    {
        tmp += nums[i] ? 1 : -1;
        if (indexes.count(tmp))
            res = max(res, i - indexes[tmp]);
        else
            indexes[tmp] = i;
    }
    return res;
}

File: 0525_contiguous_array/find_max_length.go
package main

/**
 * Time: O(n)
 * Space: O(n)
 */
func findMaxLength(nums []int) int {
	indexes := make(map[int]int)
	indexes[0] = -1
	res, tmp := 0, 0
	for i, e := range nums {
		if e == 1 {
			tmp++
		} else {
			tmp--
		}
		index, ok := indexes[tmp]
		if ok {
			res = max(res, i-index)
		} else {
			indexes[tmp] = i
		}
	}
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {

}
